core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,288 @@
+import zio.prelude._
+import zio.query.ZQuery
+import zio.{ durationInt, Duration, FiberRef, Ref, UIO, Unsafe, ZIO }
+
+import java.util.concurrent.{ ConcurrentHashMap, TimeUnit }
+
+object Caching {
+  val DirectiveName     = "cacheControl"
+  val MaxAgeName        = "maxAge"
+  val ScopeName         = "scope"
+  val InheritMaxAgeName = "inheritMaxAge"
+
+  private val cacheOverride: FiberRef[Option[CacheHint]] = Unsafe.unsafe(implicit u => FiberRef.unsafe.make(None))
+
+  private val cacheDirectives = List(
+    __Directive(
+      name = DirectiveName,
+      description = None,
+      locations = Set(__DirectiveLocation.FIELD_DEFINITION, __DirectiveLocation.OBJECT),
+      args = List(
+        __InputValue(
+          name = MaxAgeName,
+          None,
+          `type` = () => Types.int,
+          defaultValue = None
+        ),
+        __InputValue(
+          name = ScopeName,
+          None,
+          `type` = () => CacheScope.schema.toType_(),
+          defaultValue = None
+        ),
+        __InputValue(
+          name = InheritMaxAgeName,
+          None,
+          `type` = () => Types.boolean,
+          defaultValue = None
+        )
+      ),
+      isRepeatable = false
+    )
+  )
+
+  /**
+   * Computes the total cache policy for a query and stores it in the extensions of the response.
+   * The result can then be used by http adapters to set the appropriate cache headers.
+   */
+  def extension(settings: CacheSettings = CacheSettings.default): GraphQLAspect[Nothing, Any] = Default(settings)
+
+  /**
+   * Assigns a cache hint to a field or a type
+   * @param maxAge The maximum duration that this field should be cached for
+   * @param scope The scope of that should be applied to the cache
+   * @param inheritMaxAge Whether the maxAge should be inherited from the parent type
+   */
+  case class GQLCacheControl(
+    maxAge: Option[Duration] = None,
+    scope: Option[CacheScope] = None,
+    inheritMaxAge: Boolean = false
+  ) extends GQLDirective(
+        Directive(
+          DirectiveName,
+          Map(
+            MaxAgeName        -> maxAge.fold[InputValue](NullValue)(d => IntValue(d.toSeconds)),
+            ScopeName         -> scope.fold[InputValue](NullValue)(s => EnumValue(s.toString)),
+            InheritMaxAgeName -> BooleanValue(inheritMaxAge)
+          )
+        )
+      )
+
+  case class CacheSettings(
+    defaultMaxAge: Duration,
+    defaultScope: CacheScope
+  ) {
+    def withMaxAge(maxAge: Duration): CacheSettings = copy(defaultMaxAge = maxAge)
+
+    def withScope(scope: CacheScope): CacheSettings = copy(defaultScope = scope)
+  }
+
+  object CacheSettings {
+    val default: CacheSettings = CacheSettings(0.seconds, CacheScope.Public)
+  }
+
+  /**
+   * Overrides the cache hint for a particular field. This can be used to dynamically set the cache hint within
+   * a resolver
+   */
+  def setCacheHint(hint: CacheHint): UIO[Unit] = cacheOverride.set(Some(hint))
```

Comment:
Nice ❤️

Reply:
As a suggestion/question, should we check if the value was already set and use the most restrictive one instead? Something along:

```suggestion
  def setCacheHint(hint: CacheHint): UIO[Unit] =
    cacheOverride.update {
      case None           => Some(hint)
      case Some(existing) => Some(mostRestrictive(existing, hint))
    }
```

Reply:
Fixed!